lib/bolt/module_installer/puppetfile.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+require 'puppetfile-resolver'
+
+require 'bolt/error'
+require 'bolt/module_installer/puppetfile/forge_module'
+require 'bolt/module_installer/puppetfile/git_module'
+
+# This class manages the logical contents of a Puppetfile. It includes methods
+# for parsing and generating a Puppetfile.
+#
+module Bolt
+  class ModuleInstaller
+    class Puppetfile
+      attr_reader :modules
+
+      def initialize(modules = [])
+        @modules = modules
+      end
+
+      # Loads a Puppetfile and parses its modules.
+      #
+      def self.parse(path, skip_unsupported_modules: false)
+        return new unless path.exist?
+
+        begin
+          parsed = ::PuppetfileResolver::Puppetfile::Parser::R10KEval.parse(File.read(path))
+        rescue StandardError => e
+          raise Bolt::Error.new(
+            "Unable to parse Puppetfile #{path}: #{e.message}",
+            'bolt/puppetfile-parsing'
+          )
+        end
+
+        unless parsed.valid?
+          raise Bolt::ValidationError, <<~MSG
+            Unable to parse Puppetfile #{path}:
+            #{parsed.validation_errors.join("\n\n")}.
+            This may not be a Puppetfile managed by Bolt.
+          MSG
+        end
+
+        new(to_modules(parsed.modules, skip_unsupported_modules))
+      end
+
+      # Serializes the modules returned from the parser.
+      #
+      private_class_method def self.to_modules(modules, skip_unsupported)
+        modules.each_with_object([]) do |mod, acc|
+          case mod.module_type
+          when :forge
+            acc << ForgeModule.new(
+              mod.title,
+              mod.version.is_a?(String) ? mod.version[1..-1] : nil
+            )
+          when :git
+            acc << GitModule.new(
+              mod.name,
+              mod.remote,
+              mod.ref || mod.commit || mod.tag
+            )
+          else
+            unless skip_unsupported
+              raise Bolt::ValidationError,
+                    "Module '#{mod.title}' is not a Puppet Forge or Git module."
```

Comment:
```suggestion
                    "Cannot parse Puppetfile at #{path}, module '#{mod.title}' is not a Puppet Forge or Git module."
```
Or something, just to indicate what we were trying to do and where something went wrong.